### PR TITLE
Pull ros2 experimental vesc driver back into f1tenth

### PR DESCRIPTION
**Pull request by hzheng40**
The current state of the branch is tested on a car using this [container](https://hub.docker.com/layers/157411884/f1tenth/focal-l4t-foxy/vesc/images/sha256-2964ce7b21bd4f6bf1e8a47e4629bcaf4f181b5b77ce3bd62ee24578afb2a7f3?context=explore). I was able to set up the container to have NVIDIA L4T and CUDA support, with Focal 20.04 and ROS2 foxy.

A couple of modifications are still needed:
1. Setup instructions for setting up udev rules and use the rules in the params yaml
2. Port the racecar package to ROS2 and test sending commands via serial

I'm creating this PR just to get things moving.

### CODE REVIEW

File: vesc_driver/CMakeLists.txt [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
+#if(NOT "${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT}")
+#  message(STATUS "Changing CXX_STANDARD from C++98 to C++17")
+#  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
+#elseif("${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT}" STREQUAL "98")
+#  message(STATUS "Changing CXX_STANDARD from C++98 to C++17")
+#  set(CMAKE_CXX_STANDARD 17)
+#endif()
+set(CMAKE_CXX_STANDARD 17)
 find_package(ament_cmake_auto REQUIRED)
+find_package(serial_driver REQUIRED)
```
Comment (JWhitleyWork) [inline vesc_driver/CMakeLists.txt]: This is handled by `ament_auto_find_build_dependencies` below as long as the dependency is in `package.xml` so this line can be removed.

File: vesc_driver/CMakeLists.txt [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
-
+# Set minimum C++ standard to C++17
+#if(NOT "${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT}")
+#  message(STATUS "Changing CXX_STANDARD from C++98 to C++17")
+#  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
+#elseif("${CMAKE_CXX_STANDARD_COMPUTED_DEFAULT}" STREQUAL "98")
+#  message(STATUS "Changing CXX_STANDARD from C++98 to C++17")
+#  set(CMAKE_CXX_STANDARD 17)
+#endif()
+set(CMAKE_CXX_STANDARD 17)
```
Comment (JWhitleyWork) [inline vesc_driver/CMakeLists.txt]: This should use the block above that is currently commented out instead.

File: vesc_driver/include/vesc_driver/vesc_driver.hpp [hunk 3]
```diff
@@ -68,14 +67,14 @@ class VescDriver
     CommandLimit(
       rclcpp::Node * node_ptr,
       const std::string & str,
-      const boost::optional<double> & min_lower = boost::optional<double>(),
-      const boost::optional<double> & max_upper = boost::optional<double>());
+      const std::optional<double> & min_lower = std::optional<double>(),
+      const std::optional<double> & max_upper = std::optional<double>());
```
Comment (JWhitleyWork) [inline vesc_driver/include/vesc_driver/vesc_driver.hpp]: If we are going to use `std::optional` then this is missing `#include <optional>`.

File: vesc_driver/src/vesc_interface.cpp [hunk 4]
```diff
@@ -50,49 +50,70 @@ class VescInterface::Impl
...
-              ss << "Out-of-sync with VESC, unknown data leading valid frame. Discarding " <<
-                std::distance(iter_begin, iter) << " bytes.";
-              error_handler_(ss.str());
-            }
+            //if (std::distance(iter_begin, iter) > 0) {
+              //std::ostringstream ss;
+              //ss << "Out-of-sync with VESC, unknown data leading valid frame. Discarding " <<
+              //  std::distance(iter_begin, iter) << " bytes.";
+              //error_handler_(ss.str());
+            //}
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_interface.cpp]: Is there a reason this is commented out? If there is, it should be removed. If there is not, it should be un-commented.

File: vesc_driver/src/vesc_interface.cpp [hunk 5]
```diff
@@ -105,48 +126,49 @@ void VescInterface::Impl::rxThread()
             break;  // for (iter_sof...
           } else {
             // else, this was not a packet, move on to next byte
-            error_handler_(error);
+            //error_handler_(error);
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_interface.cpp]: Same here with commenting.

File: vesc_driver/src/vesc_interface.cpp [hunk 6]
```diff
@@ -105,48 +126,49 @@ void VescInterface::Impl::rxThread()
...
-        boost::asio::transfer_exactly(bytes_to_read));
-
-      if (bytes_needed > 0 && 0 == bytes_read && !buffer.empty()) {
-        error_handler_("Possibly out-of-sync with VESC, read timout in the middle of a frame.");
-      }
+      //if (std::distance(iter_begin, iter) > 0) {
+        //std::ostringstream ss;
+        //ss << "Out-of-sync with VESC, discarding " << std::distance(iter_begin, iter) << " bytes.";
+        //error_handler_(ss.str());
+      //}
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_interface.cpp]: Same here with commented code.

File: vesc_driver/src/vesc_interface_old.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
+// Copyright 2020 F1TENTH Foundation
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_interface_old.cpp]: This file should be removed. We have `git` to maintain history.

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]

**Comment by JWhitleyWork** [comment]
Hey, @hzheng40. Sorry it took me a while to get to this. There are also some merge conflicts so you will need to rebase your branch on the current `ros2` branch.

File: vesc_driver/src/vesc_interface.cpp (hunk 4 above)
Comment (HaoruXue) [reply vesc_driver/src/vesc_interface.cpp]: Hi Josh, we are working on a new version that is comparable with the old error handler. It was commented out because we cannot detect out of sync error with async receive method. We will have the revisions ready this weekend.

**Comment by HaoruXue** [comment]
Thanks @hzheng40 for getting things started! The codes right now are a bit experimental but we will clean things up and resolve the conflicts over the weekend.

File: vesc_driver/src/vesc_interface.cpp (hunk 4 above)
Comment (JWhitleyWork) [reply vesc_driver/src/vesc_interface.cpp]: Thanks! If you push the changes to this branch, the PR will automatically update. Looking forward to the updates!

**Comment by hzheng40** [comment]
@FrostXue any updates?

**Comment by JWhitleyWork** [comment]
@FrostXue Please update us ASAP. This is required for a hack-a-thon that the Foundation is hosting next week.

File: vesc_driver/CMakeLists.txt (hunk 2 above)
Comment (HaoruXue) [reply vesc_driver/CMakeLists.txt]: Hi Josh we will uncomment the above portion but we need this line to force set C++17 unless there is a better way of doing it. Removing it will cause the compiler to fall back to C++14 which is what other packages are using.

Comment (JWhitleyWork) [reply vesc_driver/CMakeLists.txt]: Please set it to C++14 and change the references to `std::optional` to `std::experimental::optional`. This should avoid the need for C++17 and be compatible with most compilers.

**Comment by HaoruXue** [comment]
Branch rebased and codes cleared up, although GitHub is still prompting "conflicting files".

Driver tested and working properly. Sensor data coming in at ~90Hz on a decent laptop.

I saw there were a couple of discussions on publishing the VESC IMU. Have the changes be pushed and do we need it for next week? If so I can work on it over the weekend.

**Comment by hzheng40** [comment]
> Driver tested and working properly. Sensor data coming in at ~90Hz on a decent laptop.

This is not the testing we are looking for. I've created a dockerfile that could be built on an actual car, testing on a laptop misses the most important part of testing: on ARM64 instead of X86_64. Also, sensor data coming one way does not mean it'll be able to listen to drive commands sent to the vesc.

**Comment by HaoruXue** [comment]
> > Driver tested and working properly. Sensor data coming in at ~90Hz on a decent laptop.
> 
> This is not the testing we are looking for. I've created a dockerfile that could be built on an actual car, testing on a laptop misses the most important part of testing: on ARM64 instead of X86_64. Also, sensor data coming one way does not mean it'll be able to listen to drive commands sent to the vesc.

Hi @hzheng40 sorry for the confusion. The test on sending control signals has been done too.

Good point on ARM vs X86. I added an additional test with an [ARM64 container](https://hub.docker.com/repository/docker/haoru233/vesc_test) on Jetson NX. Working properly. Let me know if there are anything else needed in the test.

File: vesc_driver/CMakeLists.txt (hunk 2 above)
Comment (JWhitleyWork) [reply vesc_driver/CMakeLists.txt]: This still hasn't been fixed.

File: vesc_driver/CMakeLists.txt (hunk 1 above)
Comment (JWhitleyWork) [reply vesc_driver/CMakeLists.txt]: This also still has not been fixed.

File: vesc_driver/CMakeLists.txt [hunk 8]
```diff
@@ -28,12 +28,11 @@ ament_auto_add_library(${PROJECT_NAME} SHARED
   src/vesc_packet_factory.cpp
 )
 target_link_libraries(${PROJECT_NAME}
-  ${Boost_LIBRARIES}
   ${CMAKE_THREAD_LIBS_INIT}
 )
 rclcpp_components_register_node(${PROJECT_NAME}
   PLUGIN vesc_driver::VescDriver
-  EXECUTABLE ${PROJECT_NAME}_node
+  EXECUTABLE vesc_driver_node
```
Comment (JWhitleyWork) [inline vesc_driver/CMakeLists.txt]: Why was this changed? `${PROJECT_NAME}_node` will end up producing the same text and makes it easier to maintain. Please revert.

File: vesc_driver/src/vesc_driver.cpp [hunk 9]
```diff
@@ -65,7 +65,12 @@ VescDriver::VescDriver(const rclcpp::NodeOptions & options)
   fw_version_minor_(-1)
 {
   // get vesc serial port address
-  std::string port = declare_parameter<std::string>("port", "");
+  /* default to /tmp/ttyV0 for debug purpose
+   * use :
+   *  socat /dev/ttyACM0,raw,echo=0  SYSTEM:'tee in.txt |socat - "PTY,link=/tmp/ttyV0,raw,echo=0,waitslave" |tee out.txt'
+   * to sniff comunication data to and from vesc
+   */
+  std::string port = declare_parameter<std::string>("port", "/tmp/ttyV0");
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_driver.cpp]: Changing a parameter default is usually a bad idea without a very good reason. Please remove this and use a configuration parameter for the test instead.

File: vesc_driver/src/vesc_driver.cpp [hunk 10]
```diff
@@ -171,14 +179,39 @@ void VescDriver::vescPacketCallback(const std::shared_ptr<VescPacket const> & pa
...
+    );
   }
+
+  auto & clk = *this->get_clock();
+  RCLCPP_INFO_THROTTLE(
+    get_logger(),
+    clk,
+    5000,
+    "%s packet received",
+    packet->name().c_str()
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_driver.cpp]: This is definitely a debugging message and not something that needs to be `_INFO`. Please change to `_DEBUG` or remove.

File: vesc_driver/src/vesc_interface.cpp [hunk 11]
```diff
@@ -50,49 +50,63 @@ class VescInterface::Impl
...
+    buffer_mutex_.lock();
     int bytes_needed = VescFrame::VESC_MIN_FRAME_SIZE;
-    if (!buffer.empty()) {
+    if (!buffer_.empty()) {
       // search buffer for valid packet(s)
-      Buffer::iterator iter(buffer.begin());
-      Buffer::iterator iter_begin(buffer.begin());
-      while (iter != buffer.end()) {
+      auto iter = buffer_.begin();
+      auto iter_begin = buffer_.begin();
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_interface.cpp]: This variable is no longer used in the below code. Please remove.

File: vesc_driver/src/vesc_packet.cpp [hunk 12]
```diff
@@ -34,10 +34,12 @@
 #include <iterator>
 #include <memory>
 #include <string>
+#include <cmath>
```
Comment (JWhitleyWork) [inline vesc_driver/src/vesc_packet.cpp]: I may be missing it but I don't see any functions from `cmath` being used below. Was there a reason it was added?

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]

File: vesc_driver/CMakeLists.txt (hunk 8 above)
Comment (HaoruXue) [reply vesc_driver/CMakeLists.txt]: Huh looks like the rebasing messed up some of the changes in CMakeLists. I'll get the changes back.

File: vesc_driver/src/vesc_driver.cpp (hunk 9 above)
Comment (HaoruXue) [reply vesc_driver/src/vesc_driver.cpp]: This is from commit c47fccb on the upstream branch during rebasing.

File: vesc_driver/src/vesc_driver.cpp (hunk 10 above)
Comment (HaoruXue) [reply vesc_driver/src/vesc_driver.cpp]: same from 654ba11. Could be removed.

File: vesc_driver/src/vesc_packet.cpp (hunk 12 above)
Comment (HaoruXue) [reply vesc_driver/src/vesc_packet.cpp]: from commit 654ba11

**Comment by JWhitleyWork** [comment]
@FrostXue Your rebasing is not working correctly (there are still conflicts). Do the following in your local repo to get rebased on the latest `ros2` branch from this repo:

```
git remote add upstream git@github.com:f1tenth/vesc.git
git fetch upstream
git checkout -b upstream-ros2 upstream/ros2
git checkout --track origin/ros2
git rebase upstream-ros2
... fix rebase issues and add all files with git add, then continue with git rebase --continue
... once all issues are resolved and the rebase is complete
git push -f
```

**Comment by HaoruXue** [comment]
@JWhitleyWork  Fixed. Conflcits removed.

**Review by JWhitleyWork — APPROVED** [review]

**Comment by JWhitleyWork** [comment]
@FrostXue @hzheng40 I temporarily merged this into `ros2-temp` for the hack-a-thon tomorrow. I need to get the new version of `transport_drivers` released to the build farm before we release this new version that relies on it.
